File: app/src/main/java/com/example/englishhindi/database/LessonDao.java
package com.bhashasetu.app.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.bhashasetu.app.model.Lesson;
import com.bhashasetu.app.model.LessonWord;

import java.util.List;

@Dao
public interface LessonDao {
    
    @Insert
    long insert(Lesson lesson);
    
    @Update
    void update(Lesson lesson);
    
    @Delete
    void delete(Lesson lesson);
    
    @Query("SELECT * FROM lessons ORDER BY category, orderInCategory ASC")
    LiveData<List<Lesson>> getAllLessons();
    
    @Query("SELECT * FROM lessons WHERE category = :category ORDER BY orderInCategory ASC")
    LiveData<List<Lesson>> getLessonsByCategory(String category);
    
    @Query("SELECT * FROM lessons WHERE level = :level ORDER BY category, orderInCategory ASC")
    LiveData<List<Lesson>> getLessonsByLevel(String level);
    
    @Query("SELECT * FROM lessons WHERE id = :id")
    LiveData<Lesson> getLessonById(int id);
    
    @Query("SELECT * FROM lessons WHERE hasCompleted = 0 ORDER BY level, orderInCategory ASC LIMIT 1")
    LiveData<Lesson> getNextIncompleteLesson();
    
    @Query("SELECT DISTINCT category FROM lessons ORDER BY category ASC")
    LiveData<List<String>> getAllCategories();
    
    @Query("SELECT * FROM lessons WHERE id IN " +
           "(SELECT lessonId FROM lesson_words WHERE wordId = :wordId)")
    LiveData<List<Lesson>> getLessonsContainingWord(int wordId);
    
    @Insert
    void insertLessonWord(LessonWord lessonWord);
    
    @Query("DELETE FROM lesson_words WHERE lessonId = :lessonId AND wordId = :wordId")
    void removeLessonWord(int lessonId, int wordId);
    
    @Transaction
    @Query("SELECT EXISTS(SELECT 1 FROM lesson_words WHERE lessonId = :lessonId AND wordId = :wordId)")
    boolean lessonContainsWord(int lessonId, int wordId);
}

File: app/src/main/java/com/example/englishhindi/data/dao/LessonDao.kt
package com.bhashasetu.app.data.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.bhashasetu.app.data.model.Lesson
import com.bhashasetu.app.data.model.LessonWord
import com.bhashasetu.app.data.relation.LessonWithWords
import kotlinx.coroutines.flow.Flow

/**
 * Data Access Object for the Lesson entity.
 */
@Dao
interface LessonDao {
    // Basic CRUD operations
    
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(lesson: Lesson): Long
    
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(lessons: List<Lesson>): List<Long>
    
    @Update
    suspend fun update(lesson: Lesson)
    
    @Delete
    suspend fun delete(lesson: Lesson)
    
    // LessonWord junction operations
    
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addWordToLesson(lessonWord: LessonWord)
    
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addWordsToLesson(lessonWords: List<LessonWord>)
    
    @Delete
    suspend fun removeWordFromLesson(lessonWord: LessonWord)
    
    @Query("DELETE FROM lesson_words WHERE lessonId = :lessonId AND wordId = :wordId")
    suspend fun removeWordFromLessonById(lessonId: Long, wordId: Long)
    
    @Query("DELETE FROM lesson_words WHERE lessonId = :lessonId")
    suspend fun removeAllWordsFromLesson(lessonId: Long)
    
    @Query("UPDATE lesson_words SET orderInLesson = :newOrder WHERE lessonId = :lessonId AND wordId = :wordId")
    suspend fun updateWordOrder(lessonId: Long, wordId: Long, newOrder: Int)
    
    // Queries
    
    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    suspend fun getLessonById(lessonId: Long): Lesson?
    
    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    fun getLessonByIdLiveData(lessonId: Long): LiveData<Lesson?>
    
    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    fun getLessonByIdFlow(lessonId: Long): Flow<Lesson?>
    
    @Query("SELECT * FROM lessons ORDER BY difficulty ASC, orderInCategory ASC, titleEnglish ASC")
    fun getAllLessons(): Flow<List<Lesson>>
    
    @Query("SELECT * FROM lessons ORDER BY difficulty ASC, orderInCategory ASC, titleEnglish ASC")
    fun getAllLessonsLiveData(): LiveData<List<Lesson>>
    
    @Query("SELECT * FROM lessons WHERE categoryId = :categoryId ORDER BY orderInCategory ASC, titleEnglish ASC")
    fun getLessonsByCategory(categoryId: Long): Flow<List<Lesson>>
    
    @Query("SELECT * FROM lessons WHERE difficulty = :difficulty ORDER BY orderInCategory ASC, titleEnglish ASC")
    fun getLessonsByDifficulty(difficulty: Int): Flow<List<Lesson>>
    
    @Query("SELECT * FROM lessons WHERE isCompleted = :isCompleted ORDER BY difficulty ASC, orderInCategory ASC")
    fun getLessonsByCompletionStatus(isCompleted: Boolean): Flow<List<Lesson>>
    
    @Query("SELECT * FROM lessons WHERE titleEnglish LIKE '%' || :query || '%' OR titleHindi LIKE '%' || :query || '%'")
    fun searchLessons(query: String): Flow<List<Lesson>>
    
    // Advanced queries with relationships
    
    @Transaction
    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    suspend fun getLessonWithWords(lessonId: Long): LessonWithWords?
    
    @Transaction
    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    fun getLessonWithWordsFlow(lessonId: Long): Flow<LessonWithWords?>
    
    @Transaction
    @Query("SELECT * FROM lessons")
    fun getAllLessonsWithWords(): Flow<List<LessonWithWords>>
    
    @Transaction
    @Query("SELECT * FROM lessons WHERE categoryId = :categoryId ORDER BY orderInCategory ASC")
    fun getLessonsWithWordsByCategory(categoryId: Long): Flow<List<LessonWithWords>>
    
    @Query("""
        SELECT l.* FROM lessons l
        INNER JOIN lesson_words lw ON l.id = lw.lessonId
        WHERE lw.wordId = :wordId
        ORDER BY l.difficulty ASC, l.orderInCategory ASC
    """)
    fun getLessonsContainingWord(wordId: Long): Flow<List<Lesson>>
    
    // Progress tracking
    
    @Query("UPDATE lessons SET isCompleted = :isCompleted, completedAt = :completedAt WHERE id = :lessonId")
    suspend fun updateLessonCompletion(lessonId: Long, isCompleted: Boolean, completedAt: Long?)
    
    @Query("UPDATE lessons SET lastAccessedAt = :accessedAt WHERE id = :lessonId")
    suspend fun updateLessonLastAccessed(lessonId: Long, accessedAt: Long)
    
    // Statistics queries
    
    @Query("SELECT COUNT(*) FROM lessons")
    fun getLessonCount(): Flow<Int>
    
    @Query("SELECT COUNT(*) FROM lessons WHERE isCompleted = 1")
    fun getCompletedLessonCount(): Flow<Int>
    
    @Query("SELECT COUNT(*) FROM lessons WHERE categoryId = :categoryId")
    fun getLessonCountByCategory(categoryId: Long): Flow<Int>
    
    @Query("SELECT COUNT(*) FROM lessons WHERE categoryId = :categoryId AND isCompleted = 1")
    fun getCompletedLessonCountByCategory(categoryId: Long): Flow<Int>
    
    // Utility queries
    
    @Query("SELECT EXISTS(SELECT 1 FROM lessons WHERE titleEnglish = :titleEnglish LIMIT 1)")
    suspend fun lessonExistsByTitle(titleEnglish: String): Boolean
    
    @Query("SELECT EXISTS(SELECT 1 FROM lesson_words WHERE lessonId = :lessonId AND wordId = :wordId LIMIT 1)")
    suspend fun isWordInLesson(lessonId: Long, wordId: Long): Boolean
    
    @Query("SELECT MAX(orderInCategory) + 1 FROM lessons WHERE categoryId = :categoryId")
    suspend fun getNextOrderInCategory(categoryId: Long): Int
}